bundles/org.openhab.binding.teslapowerwall/src/main/java/org/openhab/binding/teslapowerwall/internal/TeslaPowerwallWebTargets.java

```diff
@@ -0,0 +1,163 @@
+        logger.trace("getOperations response = {}", response);
+        return Operations.parse(response);
+    }
+
+    public String getToken(String email, String password)
+            throws TeslaPowerwallCommunicationException, TeslaPowerwallAuthenticationException {
+        JsonObject jsonObject = new JsonObject();
+        jsonObject.addProperty("username", "customer");
+        jsonObject.addProperty("password", password);
+        jsonObject.addProperty("email", email);
+        jsonObject.addProperty("force_sm_off", false);
+        logger.debug("logonjson = {}", jsonObject.toString());
+        String response = invoke(getTokenUri, HttpMethod.POST.asString(), "Content-Type", "application/json",
+                jsonObject.toString());
+        JsonObject jsonResponse = JsonParser.parseString(response).getAsJsonObject();
+        String token = jsonResponse.get("token").getAsString();
+        logger.debug("Token: {}", token);
+        return token;
+    }
+
+    private String invoke(String uri, String email, String password)
+            throws TeslaPowerwallCommunicationException, TeslaPowerwallAuthenticationException {
+        if (token.isEmpty()) {
+            token = getToken(email, password);
+        }
+        return invoke(uri, HttpMethod.GET.asString(), "Authorization", "Bearer " + token, "");
+    }
+
+    private String invoke(String uri, String method, String headerKey, String headerValue, String params)
+            throws TeslaPowerwallCommunicationException, TeslaPowerwallAuthenticationException {
+        logger.debug("Calling url: {}", uri);
+        int status = 0;
+        String jsonResponse = "";
+        synchronized (this) {
+            try {
+                Request request = httpClient.newRequest(uri).method(method).header(headerKey, headerValue)
+                        .timeout(TIMEOUT_MS, TimeUnit.MILLISECONDS)
+                        .content(new StringContentProvider(params), "application/json");
+                if (logger.isTraceEnabled()) {
+                    logger.trace("{} request for {}", method, uri);
+                }
+                ContentResponse response = request.send();
+                status = response.getStatus();
+                jsonResponse = response.getContentAsString();
+                if (!jsonResponse.isEmpty()) {
+                    logger.trace("JSON response: '{}'", jsonResponse);
+                }
+                if (status == HttpStatus.UNAUTHORIZED_401) {
+                    throw new TeslaPowerwallAuthenticationException("Unauthorized");
+                }
+                if (status == HttpStatus.GATEWAY_TIMEOUT_504) {
+                    throw new TeslaPowerwallCommunicationException("Gateway timeout");
+                }
+                if (!HttpStatus.isSuccess(status)) {
+                    throw new TeslaPowerwallCommunicationException(
+                            String.format("Tesla Powerwall returned error while invoking %s", uri));
```

Comment:
You could consider adding the status code here so it would be easier to backtrack what happened.

Reply:
Done.